### PR TITLE
Fix pointer of target location in visit_Associate

### LINKED ISSUE #1555: [Minpack]: Stored value type does not match pointer operand type
```fortran
subroutine find_fit(data_y)
real, intent(in) :: data_y(1)

real :: tol, fvec(1)

tol = sqrt(epsilon(1.0))

contains

    subroutine fcn(fvec)
    real :: fvec(1)
    fvec = data_y
    end subroutine fcn

end subroutine find_fit

program main

end program main
```

> LFortran
```console
$ lfortran a.f90 
; ModuleID = 'LFortran'
source_filename = "LFortran"

%array = type { float*, i32, %dimension_descriptor*, i1, i32 }
%dimension_descriptor = type { i32, i32, i32 }

@__lcompilers_created__nested_context__find_fit_data_y = global %array* null
@0 = private unnamed_addr constant [1 x i8] zeroinitializer, align 1
@c_null_ptr = external global void*

define void @find_fit(float* %data_y) {
.entry:
  %fvec = alloca %array, align 8
  %0 = getelementptr %array, %array* %fvec, i32 0, i32 1
  store i32 0, i32* %0, align 4
  %1 = getelementptr %array, %array* %fvec, i32 0, i32 2
  %2 = alloca i32, align 4
  store i32 1, i32* %2, align 4
  %3 = load i32, i32* %2, align 4
  %4 = alloca %dimension_descriptor, i32 %3, align 8
  %5 = getelementptr %array, %array* %fvec, i32 0, i32 4
  store i32 1, i32* %5, align 4
  store %dimension_descriptor* %4, %dimension_descriptor** %1, align 8
  %6 = load %dimension_descriptor*, %dimension_descriptor** %1, align 8
  %7 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %6, i32 0
  %8 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 0
  %9 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 1
  %10 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 2
  store i32 1, i32* %8, align 4
  store i32 1, i32* %9, align 4
  store i32 1, i32* %10, align 4
  %11 = alloca i32, align 4
  store i32 1, i32* %11, align 4
  %12 = getelementptr %array, %array* %fvec, i32 0, i32 0
  %13 = load i32, i32* %11, align 4
  %14 = alloca float, i32 %13, align 4
  store float* %14, float** %12, align 8
  %tol = alloca float, align 4
  store float* %data_y, %array** @__lcompilers_created__nested_context__find_fit_data_y, align 8
  store float 0x3F36A09E60000000, float* %tol, align 4
  br label %return

return:                                           ; preds = %.entry
  ret void
}

declare i8* @__module_lfortran_intrinsic_builtin_char(i32*)

define float @__module_lfortran_intrinsic_math_sepsilon(float* %x) {
.entry:
  %r = alloca float, align 4
  store float 0x3E80000000000000, float* %r, align 4
  br label %return

return:                                           ; preds = %.entry
  %0 = load float, float* %r, align 4
  ret float %0
}

define float @__module_lfortran_intrinsic_math_ssqrt(float* %x) {
.entry:
  %r = alloca float, align 4
  %0 = load float, float* %x, align 4
  %1 = call float @llvm.sqrt.f32(float %0)
  store float %1, float* %r, align 4
  br label %return

return:                                           ; preds = %.entry
  %2 = load float, float* %r, align 4
  ret float %2
}

; Function Attrs: nounwind readnone speculatable willreturn
declare float @llvm.sqrt.f32(float) #0

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_set_argv(i32 %0, i8** %1)
  ret i32 0
}

declare void @_lpython_set_argv(i32, i8**)

attributes #0 = { nounwind readnone speculatable willreturn }
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store float* %data_y, %array** @__lcompilers_created__nested_context__find_fit_data_y, align 8
 %array*


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that must be fixed).
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1555 
With this PR, `minpack/example_primes.f90` successfully compile to LLVM.

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -4259,7 +4261,36 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                ASRUtils::check_equal_type(target_type, value_type) ) {
+                llvm::Value* llvm_target_ptr = llvm_target;
+                llvm::Value* llvm_value_ptr = llvm_value;
+
+                llvm::Value* target_array = llvm_target;
+                llvm::Value* value_array = llvm_value;
+
+                llvm::Value* target_ptr = llvm_target;
+                llvm::Value* value_ptr = llvm_value;
+                if (llvm_value->getType()->isPointerTy() && llvm_value->getType()->getPointerElementType()->isFloatTy()) {
```
Comment (Pranavchiku) [inline src/libasr/codegen/asr_to_llvm.cpp:4275]: I am not sure if using this is fine.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -4259,7 +4261,39 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                llvm::Value* target_ptr = llvm_target;
+                llvm::Value* value_ptr = llvm_value;
+
+                std::vector<llvm::Value*> idx_vec = {
+                    llvm::ConstantInt::get(context, llvm::APInt(32, 0))};
+                if (llvm_value->getType()->isPointerTy() && ( llvm_value->getType()->getPointerElementType()->isFloatTy() || llvm_value->getType()->getPointerElementType()->isDoubleTy() )) {
+                    // check if target is not float*
+                    if (!(llvm_target->getType()->isPointerTy() && ( llvm_target->getType()->getPointerElementType()->isFloatTy() || llvm_target->getType()->getPointerElementType()->isDoubleTy() ))) {
+                        llvm_target_ptr = CreateLoad(llvm_target);
+                        target_array = arr_descr->get_pointer_to_data(llvm_target_ptr);
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:4282]: `llvm_target_ptr` is always NULL from the start. `llvm_target` doesn't point to anything. So first we will have to check if its `NULL`, if so then do `alloca` for it. Once `alloca` is done then you can use it store data pointer. Also, once we need to fill the allocated descriptor with the dimensional details from ASR types.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -4259,7 +4261,39 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                        target_ptr = CreateGEP(target_array, idx_vec);
+                    }
+                } else {
+                    // check if target is float*
+                    if (llvm_target->getType()->isPointerTy() && ( llvm_target->getType()->getPointerElementType()->isFloatTy() || llvm_target->getType()->getPointerElementType()->isDoubleTy() )) {
+                        llvm_value_ptr = CreateLoad(llvm_value);
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:4288]: This is not correct. We should not use LLVM type APIs when ASR has the desired information.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Comment by czgdp1807** [comment]
I have some time so let me try it out.

**Comment by Pranavchiku** [comment]
> I have some time so let me try it out.

Perfect. Till then I will use this PR as it compiles example_primes to LLVM to discover and resolve issues.

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/lfortran/lfortran/pull/1574

**Comment by czgdp1807** [comment]
Both the tests in this PR work fine with `main` branch of LFortran as well. So they aren't as useful as the one in https://github.com/lfortran/lfortran/issues/1555#issue-1667561589 (i.e., the issue itself).

**Comment by Pranavchiku** [comment]
That example works as well. I tried to create meaningful function calls in that, but I was unable to do so. Hence I used these.

**Pull request closed without merging**